File: testing/compose/js/chatbot/src/skills/catalogue.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";

import type { ToolName } from "../tools/catalogue.js";

export interface SkillFixture {
  skill: string;
  command: string;
  mock_script: Array<{ tool: ToolName; stub_output: Record<string, unknown> }>;
  expected_event: {
    task: { category: string; sub_category: string; outcome: string };
    metrics: Record<string, number>;
    context: Record<string, unknown>;
  };
}

const SKILL_ALIASES: Record<string, string> = {
  "code-gen": "code_gen",
  "doc-summary": "doc_summary",
  "ci-triage": "ci_triage",
};

export function normalizeSkill(raw: string | undefined): string {
  const value = (raw ?? "code_gen").replace(/^\/skill\s+/, "").trim();
  return SKILL_ALIASES[value] ?? value;
}

export function loadSkill(raw: string | undefined): SkillFixture {
  const skill = normalizeSkill(raw);
  const fixtureRoot = process.env["SKILL_FIXTURE_DIR"] ?? "/fixtures/skills";
  const fixturePath = join(fixtureRoot, `${skill}.json`);
  return JSON.parse(readFileSync(fixturePath, "utf8")) as SkillFixture;
}

export function activeTools(fixture: SkillFixture): ToolName[] {
  return fixture.mock_script.map((step) => step.tool);
}

File: testing/compose/rs/chatbot/src/skills/catalogue.rs
use serde::Deserialize;
use std::path::PathBuf;

#[derive(Debug, Clone, Deserialize)]
pub struct SkillFixture {
    pub skill: String,
    pub mock_script: Vec<MockScriptStep>,
    pub expected_event: ExpectedEvent,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MockScriptStep {
    pub tool: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ExpectedEvent {
    pub task: serde_json::Value,
    pub metrics: serde_json::Value,
    pub context: serde_json::Value,
}

pub fn normalize_skill(raw_skill: Option<&str>) -> String {
    let selected = raw_skill
        .unwrap_or("code_gen")
        .trim()
        .strip_prefix("/skill ")
        .unwrap_or_else(|| raw_skill.unwrap_or("code_gen").trim())
        .to_owned();
    match selected.as_str() {
        "code-gen" => "code_gen".to_owned(),
        "doc-summary" => "doc_summary".to_owned(),
        "ci-triage" => "ci_triage".to_owned(),
        "" => "code_gen".to_owned(),
        _ => selected,
    }
}

pub fn load_skill(raw_skill: Option<&str>) -> anyhow::Result<SkillFixture> {
    let fixture_root =
        std::env::var("SKILL_FIXTURE_DIR").unwrap_or_else(|_| "/fixtures/skills".to_owned());
    let mut fixture_path = PathBuf::from(fixture_root);
    fixture_path.push(format!("{}.json", normalize_skill(raw_skill)));
    let body = std::fs::read_to_string(&fixture_path)?;
    Ok(serde_json::from_str(&body)?)
}

pub fn active_tools(fixture: &SkillFixture) -> Vec<String> {
    fixture
        .mock_script
        .iter()
        .map(|step| step.tool.clone())
        .collect()
}
